cpp/ql/src/Security/CWE/CWE-311/CleartextBufferWrite.ql

```diff
@@ -19,7 +19,30 @@ import semmle.code.cpp.ir.dataflow.TaintTracking
 import DataFlow::PathGraph
 
 /**
- * A taint flow configuration for flow from user input to a buffer write.
+ * A buffer write into a sensitive expression.
+ */
+class SensitiveBufferWrite extends Expr {
+  BufferWrite::BufferWrite write;
+
+  SensitiveBufferWrite() {
+    this = write and
+    write.getDest() instanceof SensitiveExpr
+  }
+
+  /**
+   * Gets a data source of this operation.
+   */
+  Expr getASource() { result = write.getASource() }
+
+  /**
+   * Gets the destination buffer of this operation.
+   */
+  Expr getDest() { result = write.getDest() }
+}
```

Comment:
The `this = foo` pattern in a charpred can always be rewritten into a non-extending subtype relation. This will be easier for the compiler to optimize.
```suggestion
class SensitiveBufferWrite extends Expr instanceof BufferWrite::BufferWrite {
  SensitiveBufferWrite() { super.getDest() instanceof SensitiveExpr }

  /**
   * Gets a data source of this operation.
   */
  Expr getASource() { result = super.getASource() }

  /**
   * Gets the destination buffer of this operation.
   */
  Expr getDest() { result = super.getDest() }
}
```